File: backend/Server/Services/MenuAddonsService.cs
using backend.Server.Database;
using backend.Server.Interfaces;
using backend.Server.Models.DatabaseObjects;
using backend.Server.Models.DTOs.MenuAddon;
using backend.Server.Exceptions;
using backend.Server._helpers;
using Microsoft.EntityFrameworkCore;

namespace backend.Server.Services
{
    public class MenuAddonsService(ApplicationDbContext context) : IMenuAddonsService
    {
        private readonly ApplicationDbContext _context = context;

        public async Task<List<MenuItemIngredient>> GetAllMenuAddonsAsync(MenuAddonsGetAllDTO request)
        {
            if (request.Page < 0)
            {
                throw new ApiException(400, "Page number must be greater than zero");
            }
            if (request.PerPage <= 0)
            {
                throw new ApiException(400, "PerPage value must be greater than zero");
            }

            if (request.Page == 0)
            {
                return await _context.MenuItemIngredients
                    .AsNoTracking()
                    .ToListAsync();
            }

            return await _context.MenuItemIngredients
                .Skip((request.Page - 1) * request.PerPage)
                .Take(request.PerPage)
                .AsNoTracking()
                .ToListAsync();
        }

        public async Task<MenuItemIngredient> CreateMenuAddonAsync(MenuAddonCreateDTO request)
        {
            if (await _context.MenuItemIngredients.AnyAsync(m => m.Name == request.Name && m.GroupId == request.GroupId))
            {
                throw new ApiException(409, $"Menu addon with the same name {request.Name} already exists in this group {request.GroupId}.");
            }

            // Check if the ingredient group exists
            var group = await _context.MenuItemIngredientGroups.FindAsync(request.GroupId);
            if (group == null)
            {
                throw new ApiException(404, $"Ingredient group {request.GroupId} not found");
            }

            var menuAddon = new MenuItemIngredient
            {
                Name = request.Name,
                GroupId = request.GroupId,
                Price = request.Price
            };

            _context.MenuItemIngredients.Add(menuAddon);

            await Helper.SaveChangesOrThrowAsync(_context, "Failed to create menu addon");

            return menuAddon;
        }

        public async Task<MenuItemIngredient> GetMenuAddonByNidAsync(long nid)
        {
            if (nid <= 0)
            {
                throw new ApiException(400, "Invalid menu addon Nid");
            }

            var menuAddon = await _context.MenuItemIngredients.FindAsync(nid) ?? throw new ApiException(404, $"Menu addon {nid} not found");
            
            return menuAddon;
        }

        public async Task<List<MenuItemIngredient>> GetMenuAddonsByMenuItemNidAsync(long menuItemNid)
        {
            if (menuItemNid <= 0)
            {
                throw new ApiException(400, "Invalid menu item ID");
            }

            // Get all groups for this menu item, then get all addons from those groups
            var groups = await _context.MenuItemIngredientGroups
                .Where(g => g.MenuItemId == menuItemNid)
                .Select(g => g.Nid)
                .ToListAsync();

            var addons = await _context.MenuItemIngredients
                .Where(m => m.GroupId != null && groups.Contains(m.GroupId.Value))
                .AsNoTracking()
                .ToListAsync();

            return addons;
        }

        public async Task<List<MenuItemIngredient>> GetMenuAddonsByGroupNidAsync(long groupNid)
        {
            if (groupNid <= 0)
            {
                throw new ApiException(400, "Invalid group ID");
            }

            var addons = await _context.MenuItemIngredients
                .Where(m => m.GroupId == groupNid)
                .AsNoTracking()
                .ToListAsync();

            return addons;
        }

        public async Task UpdateMenuAddonAsync(MenuAddonUpdateDTO request, long nid)
        {
            if (nid <= 0)
            {
                throw new ApiException(400, "Invalid menu addon Nid");
            }
            var menuAddon = await _context.MenuItemIngredients.FindAsync(nid) ?? throw new ApiException(404, $"Menu addon {nid} not found");

            if (request.Name != null) menuAddon.Name = request.Name;
            if (request.GroupId.HasValue)
            {
                // Verify the group exists
                var group = await _context.MenuItemIngredientGroups.FindAsync(request.GroupId.Value);
                if (group == null)
                {
                    throw new ApiException(404, $"Ingredient group {request.GroupId.Value} not found");
                }
                menuAddon.GroupId = request.GroupId.Value;
            }
            if (request.Price.HasValue) menuAddon.Price = request.Price.Value;

            _context.MenuItemIngredients.Update(menuAddon);
            
            await Helper.SaveChangesOrThrowAsync(_context, "Failed to update menu addon", expectChanges: false);
        }

        public async Task DeleteMenuAddonAsync(long nid)
        {
            if (nid <= 0)
            {
                throw new ApiException(400, "Invalid menu addon Nid");
            }
            var menuAddon = await _context.MenuItemIngredients.FindAsync(nid) ?? throw new ApiException(404, $"Menu addon {nid} not found");

            _context.MenuItemIngredients.Remove(menuAddon);

            await Helper.SaveChangesOrThrowAsync(_context, "Failed to delete menu addon");
        }
    }
}

File: frontend/src/services/menuAddonsService.ts
const API_BASE_URL = 'http://localhost:5087/api';

export interface MenuItemIngredient {
  nid: number;
  groupId?: number;
  name: string;
  price: number;
}

export interface MenuAddonCreateDTO {
  name: string;
  groupId: number;
  price: number;
}

export interface MenuAddonUpdateDTO {
  name?: string;
  groupId?: number;
  price?: number;
}

export interface MenuAddonGroup {
  nid: number;
  name: string;
  menuItemId: number;
}

export interface MenuAddonGroupCreateDTO {
  name: string;
  menuItemId: number;
}

export interface MenuAddonGroupUpdateDTO {
  name?: string;
  menuItemId?: number;
}

export const menuAddonsApi = {
  // GET /api/menu/addons?Page=0&PerPage=10
  getAllAddons: async (page: number = 0, perPage: number = 100): Promise<MenuItemIngredient[]> => {
    const response = await fetch(
      `${API_BASE_URL}/menu/addons?Page=${page}&PerPage=${perPage}`
    );
    if (!response.ok) throw new Error('Failed to fetch addons');
    return await response.json();
  },

  // GET /api/menu/addons/{nid}
  getAddonByNid: async (nid: number): Promise<MenuItemIngredient> => {
    const response = await fetch(`${API_BASE_URL}/menu/addons/${nid}`);
    if (!response.ok) throw new Error('Failed to fetch addon');
    return await response.json();
  },

  // GET /api/menu/addons/by-group/{groupNid}
  getAddonsByGroupNid: async (groupNid: number): Promise<MenuItemIngredient[]> => {
    const response = await fetch(`${API_BASE_URL}/menu/addons/by-group/${groupNid}`);
    if (!response.ok) throw new Error('Failed to fetch addons by group');
    return await response.json();
  },

  // POST /api/menu/addons
  createAddon: async (data: MenuAddonCreateDTO): Promise<MenuItemIngredient> => {
    const response = await fetch(`${API_BASE_URL}/menu/addons`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    if (!response.ok) throw new Error('Failed to create addon');
    return await response.json();
  },

  // PUT /api/menu/addons/{nid}
  updateAddon: async (nid: number, data: MenuAddonUpdateDTO): Promise<void> => {
    const response = await fetch(`${API_BASE_URL}/menu/addons/${nid}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    if (!response.ok) throw new Error('Failed to update addon');
  },

  // DELETE /api/menu/addons/{nid}
  deleteAddon: async (nid: number): Promise<void> => {
    const response = await fetch(`${API_BASE_URL}/menu/addons/${nid}`, {
      method: 'DELETE',
    });
    if (!response.ok) throw new Error('Failed to delete addon');
  },
};

export const addonGroupsApi = {
  // GET /api/menu/addon-groups?Page=0&PerPage=10
  getAllGroups: async (page: number = 0, perPage: number = 100): Promise<MenuAddonGroup[]> => {
    const response = await fetch(
      `${API_BASE_URL}/menu/addon-groups?Page=${page}&PerPage=${perPage}`
    );
    if (!response.ok) throw new Error('Failed to fetch addon groups');
    return await response.json();
  },

  // GET /api/menu/addon-groups/{nid}
  getGroupByNid: async (nid: number): Promise<MenuAddonGroup> => {
    const response = await fetch(`${API_BASE_URL}/menu/addon-groups/${nid}`);
    if (!response.ok) throw new Error('Failed to fetch addon group');
    return await response.json();
  },

  // GET /api/menu/addon-groups/by-menu-item/{menuItemNid}
  getGroupsByMenuItemNid: async (menuItemNid: number): Promise<MenuAddonGroup[]> => {
    const response = await fetch(`${API_BASE_URL}/menu/addon-groups/by-menu-item/${menuItemNid}`);
    if (!response.ok) throw new Error('Failed to fetch addon groups by menu item');
    return await response.json();
  },

  // POST /api/menu/addon-groups
  createGroup: async (data: MenuAddonGroupCreateDTO): Promise<MenuAddonGroup> => {
    const response = await fetch(`${API_BASE_URL}/menu/addon-groups`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    if (!response.ok) throw new Error('Failed to create addon group');
    return await response.json();
  },

  // PUT /api/menu/addon-groups/{nid}
  updateGroup: async (nid: number, data: MenuAddonGroupUpdateDTO): Promise<void> => {
    const response = await fetch(`${API_BASE_URL}/menu/addon-groups/${nid}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    if (!response.ok) throw new Error('Failed to update addon group');
  },

  // DELETE /api/menu/addon-groups/{nid}
  deleteGroup: async (nid: number): Promise<void> => {
    const response = await fetch(`${API_BASE_URL}/menu/addon-groups/${nid}`, {
      method: 'DELETE',
    });
    if (!response.ok) throw new Error('Failed to delete addon group');
  },
};
